---
title: "Aggregate loss of function variants, wecare feb2016"
author: "Alexey Larionov"
output: html_document
---

started: Mar2016
last_updated:26Mar2016

# start_section

```{r start_section}

# Start time
Sys.time()

# Clean-up
rm(list=ls())
graphics.off()

# Set root working folder
library(knitr)
opts_knit$set(root.dir = "/scratch/medgen/users/alexey/wecare_feb2016/wecare_feb2016_r")
#setwd("/scratch/medgen/users/alexey/wecare_feb2016/wecare_feb2016_r")

```

# load_and_check_data

```{r load_and_check_data}

load(file="data/s05_add_subgroups_afs_feb2016.RData")
ls()

dim(genotypes.mx)
genotypes.mx[1:5,1:5]

dim(phenotypes.df)
colnames(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:5,1:5]

dim(variants.df)
colnames(variants.df)
str(variants.df)
variants.df[1:5,1:5]

# Check consistence of tables
sum(rownames(genotypes.mx) != rownames(variants.df))
sum(colnames(genotypes.mx) != rownames(phenotypes.df))

```

# prepare_genes_list

```{r prepare_genes_list}

# Get list of genes
genes <- as.vector(variants.df[ ,"SYMBOL"] )
length(genes)

x <- sort(table(genes), decreasing = TRUE)
x[1:86] # >3 types of lof events per gene (~2% genes with lof events)
x[87:254] # 3 types of lof events per gene (168 genes, ~4% genes with lof events)
x[255:927] # 2 types of lof events per gene (673 genes, ~17% genes with lof events)

# Visual inspection of the genes with multiple types of lof events
#-----------------------------------------------------------------

# It is immediately noticeable that ATM is at the very top of the list: 8 different lof events in wecare. 
# The genes with multiple types of lof events also included: BRCA1 (5 types of events), 
# BRCA2 (4 types of lof events), PALB2 (3 types of lof events) and RAD52 (2 types of lof events), 

# This suggests that diversity of lof events per gene (= number of diferent lof events per gene) 
# may be a meaningful feature, along with the total events burden, of cource.  

# Interestingly, CYP2D6 had 4 types of lof events and MKI67 had 2 types of lof events.  

# Also it could be noted that a number of collagens and MMPs showed multiple (2-3) types of lof events.  
# This could be mechanistically linked to extracellular matrix in tumours or normal breast.  

hist(x, breaks=c(-0.5:14.5), xlim=c(0,14), ylim=c(0,3500), 
     labels = TRUE, main="HI-LOF variants per gene", 
     xlab="Num of HI-LOF variants", ylab="Genes count", 
     xaxp=c(0,14,14))

rm(x)

# Make list of unique genes names
genes <- unique(genes)
length(genes)

```

# aggregate_genotypes_table

```{r aggregate_genotypes_table}

# Prepare header
events.per.gene.df <- c("SYMBOL",colnames(genotypes.mx))

# Aggregate GT data by genes 
# (NB: variants.df and genotypes.df have the same order of variants!)
for (gene in genes){
  
  variants <- variants.df[ ,"SYMBOL"] == gene
  variants.count <- sum(variants, na.rm=TRUE)
  gene.data <- genotypes.mx[ variants , ]

  if (variants.count == 1) {
    aggregate <- gene.data
  } else {
    aggregate <- apply(gene.data, 2, sum, na.rm=TRUE)
  }

  events.per.gene.df <- rbind( events.per.gene.df , c(gene,aggregate))

}

# Update colnames
colnames(events.per.gene.df) <- events.per.gene.df[1,]
events.per.gene.df <- events.per.gene.df[-1,]

# Update rownames
rownames(events.per.gene.df) <- events.per.gene.df[,1]
events.per.gene.df <- events.per.gene.df[,-1]

# Convert to numeric
genes <- rownames(events.per.gene.df)
events.per.gene.mx <- apply(events.per.gene.df,2,as.numeric)
genes -> rownames(events.per.gene.mx)

# Check dimentions
dim(events.per.gene.mx)
events.per.gene.mx[1:5,1:5]

# Check the summation
sum(genotypes.mx, na.rm = TRUE)
sum(events.per.gene.mx, na.rm=TRUE)

# Clean-up
rm(gene, variants, variants.count, gene.data, aggregate, genotypes.mx, events.per.gene.df)

```

# aggregate_variants_table

```{r aggregate_variants_table}

# Check that the required columns are numeric
str(variants.df[, c("AF", "AC", "AN", 
  "AF.cbc", "AC.cbc", "AN.cbc", "AF.ubc", "AC.ubc", "AN.ubc", 
  "AF.cfbc", "AC.cfbc", "AN.cfbc", "AF.usbc", "AC.usbc", "AN.usbc")])

# Prepare header for the output table
burden.per.gene.df <- c("SYMBOL", "BD", "EC", "EN", 
  "BD.cbc", "EC.cbc", "EN.cbc", "BD.ubc", "EC.ubc", "EN.ubc", 
  "BD.cfbc", "EC.cfbc", "EN.cfbc", "BD.usbc", "EC.usbc", "EN.usbc")

# EB stands for Events Burden (may be changed to Events Load, 
# if "burden" has a special meaning in statistical context).
# Importantly, despite being a sum of frequencies, it is not 
# a frequency and it can be > 1

# EC and EN stand for Events Count/Number, by analogy with AF/AN
# (AF/AN stand for Alleles Count/Number in VCF files)

# Aggregate AF, AN and AC by genes (for each group of samples)
for (gene in genes){
  
  variants <- variants.df[ ,"SYMBOL"] == gene
  variants.count <- sum(variants, na.rm=TRUE)
  gene.data <- variants.df[variants, c("AF", "AC", "AN", 
  "AF.cbc", "AC.cbc", "AN.cbc", "AF.ubc", "AC.ubc", "AN.ubc", 
  "AF.cfbc", "AC.cfbc", "AN.cfbc", "AF.usbc", "AC.usbc", "AN.usbc")]

  if (variants.count == 1) {
    aggregate <- gene.data
  } else {
    aggregate <- apply(gene.data, 2, sum, na.rm = TRUE)
  }

  burden.per.gene.df <- rbind(burden.per.gene.df , c(gene,aggregate))

}

# Update colnames
colnames(burden.per.gene.df) <- burden.per.gene.df[1,]
burden.per.gene.df <- burden.per.gene.df[-1,]

# Update rownames
rownames(burden.per.gene.df) <- burden.per.gene.df[,1]
burden.per.gene.df <- burden.per.gene.df[,-1]

# Convert to numeric
genes <- rownames(burden.per.gene.df)
burden.per.gene.mx <- apply(burden.per.gene.df,2,as.numeric)
genes -> rownames(burden.per.gene.mx)

# Check dimentions
dim(burden.per.gene.mx)

# Check the summation
sum(variants.df[, c("AF", "AC", "AN", 
  "AF.cbc", "AC.cbc", "AN.cbc", "AF.ubc", "AC.ubc", "AN.ubc", 
  "AF.cfbc", "AC.cfbc", "AN.cfbc", "AF.usbc", "AC.usbc", "AN.usbc")],
  na.rm = TRUE)

sum(burden.per.gene.mx)

# Clean-up
rm(gene, genes, variants, variants.count, gene.data, aggregate, burden.per.gene.df)

```

# explore_results

```{r explore_results}

# Explore selected genes 
# Note that EFs per gene may exceed 1.
# This means that there could be several events per one subject
# i.e. two or three LOF variants in the same sample. 
genes_list <- c("NPHP4","OR4C5","SARM1","ATM","PALB2", "MKI67")
genes_index <- rownames(burden.per.gene.mx) %in% genes_list
burden.per.gene.mx[genes_index,]
rm(genes_list, genes_index)

```

# save_data

```{r save_data}

save.image(file="data/s06_aggregate_lof_feb2016.RData")

```

# final_section

```{r final_section}

sessionInfo()
Sys.time()
  
```
