---
title: "Reshape and clean data: wecare feb2016"
author: "Alexey Larionov"
output: html_document
---

started: 01Mar2016
last updated: 15May2016

# start_section

```{r start_section}

# Start time
Sys.time()

# Clean-up
rm(list=ls())
graphics.off()

# Set root working folder
library(knitr)
opts_knit$set(root.dir = "/scratch/medgen/scripts/rscripts_05.16")
#setwd("/scratch/medgen/scripts/rscripts_05.16")

#opts_knit$set(root.dir = "C:\\Users\\larion01\\Documents\\GitHub\\R1")
#setwd("C:\\Users\\larion01\\Documents\\GitHub\\R1")

# Required libraries
library(tidyr) # for separate
library(dplyr) # for piping, filter, select etc
library(stringr) # for str_replace_all

```

# load_data

```{r load_data}

load(file="data/s04_filter_by_effect_feb2016.RData")
ls()

vv.df <- vv.std
gt.mx <- gt.std

rm(vv.std, gt.std, gq.std, dp.std, ref.std, alt.std)
rm(vv.str, gt.str, gq.str, dp.str, ref.str, alt.str)
rm(vv.rel, gt.rel, gq.rel, dp.rel, ref.rel, alt.rel)

ls()

# I do NOT convert_dfs_to_tables because it loses row manes. 
# However, it could be considered later, 
# when row names handling in tables is fixed. 

```

# check_data

```{r check_data}

dim(covar.df)
str(covar.df)
covar.df[1:5,1:5]

dim(demographics.df)
str(demographics.df)
demographics.df[1:5,1:5]

dim(samples.df)
str(samples.df)
samples.df[1:5,]

dim(vv.df)
str(vv.df)
vv.df[1:5,1:5]

dim(gt.mx)
class(gt.mx)
gt.mx[10:25,1:5]

# Check consistence of rownames
sum(rownames(gt.mx) != rownames(vv.df), na.rm=TRUE)

```

# reshape_variants_annotations

```{r reshape_variants_annotations}

# select relevant variants annotations
variants.df <- 
  vv.df %>% 
  select(SYMBOL, TYPE, CHROM, POS, REF, ALT, AC, AF, AN, 
         Consequence, SIFT, PolyPhen, CLIN_SIG, 
         GMAF, EUR_MAF, ALT_frequency_in_1k_95)

rm(vv.df)

# Explore cases with confusing GMAF/EUR_MAF containing & symbol
nrow(variants.df %>% filter(grepl("&",GMAF)))
nrow(variants.df %>% filter(grepl("&",EUR_MAF)))
variants.df %>% 
  filter(grepl("&",EUR_MAF)) %>% 
  select(c(1:6), GMAF, EUR_MAF) %>% 
  top_n(10,EUR_MAF)

# There are even 3 variants with triplicated values separated by &
variants.df[c(1235, 11405, 15894),c("GMAF", "EUR_MAF")]

# Split cases with confusing EUR_MAF containing "&"" symbol
variants.df <- 
  variants.df %>% 
  separate(EUR_MAF, c("EUR_MAF1", "EUR_MAF2"), "&") 

# All duplicated values are the same
sum(variants.df$EUR_MAF1 != variants.df$EUR_MAF2, na.rm=TRUE)

# Remove one of the duplicates
variants.df <- 
  variants.df %>% 
  select(-EUR_MAF2)

# split GMAF and EUR_MAF1
variants.df <- 
  variants.df %>% 
  separate(GMAF, c("GMAF_Allele", "GMAF_Fraction"), ":") %>% 
  separate(EUR_MAF1, c("EUR_MAF_Allele", "EUR_MAF_Fraction"), ":") 

# Convert columns to numeric etc
as.numeric(variants.df$GMAF_Fraction) -> variants.df$GMAF_Fraction
as.numeric(variants.df$EUR_MAF_Fraction) -> variants.df$EUR_MAF_Fraction

# Change "" to NAs
NA -> variants.df$GMAF_Allele[variants.df$GMAF_Allele == ""]
NA -> variants.df$EUR_MAF_Allele[variants.df$EUR_MAF_Allele == ""]
NA -> variants.df$SIFT[variants.df$SIFT == ""]
NA -> variants.df$PolyPhen[variants.df$PolyPhen == ""]
NA -> variants.df$CLIN_SIG[variants.df$CLIN_SIG == ""]

# Convert chars to factors
as.factor(variants.df$GMAF_Allele) -> variants.df$GMAF_Allele
as.factor(variants.df$EUR_MAF_Allele) -> variants.df$EUR_MAF_Allele

# Change "true" to TRUE
sum(variants.df$ALT_frequency_in_1k_95 == "true", na.rm = TRUE)
variants.df <- 
  variants.df %>% 
  mutate(frequent_in_1k=as.logical(
    str_replace_all(ALT_frequency_in_1k_95, "true", TRUE))) %>% 
  select(-ALT_frequency_in_1k_95)
sum(variants.df$frequent_in_1k, na.rm=TRUE)

# Split SIFT
variants.df <- 
  variants.df %>% 
  mutate(SIFT_call=sub("\\(.*\\)","",SIFT)) %>% 
  mutate(SIFT_score=as.numeric(
    sub(".*\\(","", sub("\\)","",SIFT)))) %>% 
  select(-SIFT)

# Split PolyPhen
variants.df <- 
  variants.df %>% 
  mutate(PolyPhen_call=sub("\\(.*\\)","",PolyPhen)) %>% 
  mutate(PolyPhen_score=as.numeric(
    sub(".*\\(","", sub("\\)","",PolyPhen)))) %>% 
  select(-PolyPhen)

# Check variants table
dim(variants.df)
str(variants.df)
variants.df[1:5,1:5]

```

# reshape_covar

```{r reshape_covar}

dim(covar.df)
colnames(covar.df)

attach(covar.df)

# chemo is the same as chemo_self_mra, no missed values
sum(chemo != chemo_self_mra)

# One missed case in hormone/ hormone_self_mra
sum(is.na(hormone))
sum(is.na(hormone_self_mra))

# hormone has 10 differences from hormone_self_mra
sum(hormone != hormone_self_mra, na.rm=TRUE)

# status is the same as status2, no missed values
sum(status != status2)

# Race is always 0, no missed data
sum(race)

# Dismiss split 495 vs 3, no missed data
sum(dsmiss)

# good_location split 86 vs 412, no missed data
sum(good_location)

# Deleterious split 480 vs 18, no missed data
sum(Deleterious)

# Deleterious split 158 vs 340, no missed data
sum(family_history)

# XRTBrCHAR is the same as XRTBreast, no missed values
sum(XRTBrCHAR != XRTBreast)

# sub_dx_age_cat split 169 vs 329, no missed data
sum(sub_dx_age_cat)

# dose
summary(dose)

detach(covar.df)

# Keep only selected annotations
selected_annotations <- c(
  "labid","cc","sub_dx_age","offset",
  "chemo","hormone","XRTBreast","dose",
  "Eigen_1","Eigen_2","Eigen_3","Eigen_4","Eigen_5")

# Keep only selected annotations
covar.df <- covar.df[,selected_annotations]
rm(selected_annotations)

# change XRTBreast to xrt
"xrt" -> colnames(covar.df)[7]

# Check result
dim(covar.df)
str(covar.df)
covar.df[1:5,1:5]


```

# explore_demographics

```{r explore_demographics}

# Outlook
dim(demographics.df)
colnames(demographics.df)
demographics.df[1:10,1:5]

# Exclude odd cases
odd_cases=c("201558","201921","202698","21IAno","387460",
            "389192","58IR1+","60IA1+","92IRno","98IAno")

demographics.df %>% 
  filter(Subject_ID %in% odd_cases) %>% 
  select(1:5)

pheno.df <- demographics.df %>% filter(! Subject_ID %in% odd_cases)
dim(pheno.df)
pheno.df[1:10,1:5]

# Explore non-odd cases
attach(pheno.df)

# --- Compare "x" and "y" columns --- #

# Identical columns
ID.x[1:5]
sum(ID.x != ID.y)

cc.x[1:20]
sum(cc.x != cc.y)

labid.x[1:5]
sum(as.vector(labid.x) != as.vector(labid.y))

fam_hist[1:5]
sum(fam_hist != family_history.y)
sum(as.numeric(sub("none",0,sub("1\\+",1,family_history.x))) != family_history.y)

sub_dx_age.x[1:5]
sum(sub_dx_age.x != sub_dx_age.y)

hist(sub_dx_age.x) 
# Consistent with 55 yr selection criteria

dxyr_stratum[1:5]
age_stratum1[1:5]
sub_dx_age_cat[1:5]

sub_dx_age.x[1:5]
refage[1:5]

offset.x[1:5]
sum(offset.x != offset.y)

hist(offset.x) 
# consistent with 1+ year between the 1st and 2nd tumours (selection criteria)

chemo_hormone[1:5]

treatment.x[1:5]
treatment.y[1:5]
sum(treatment.x != treatment.y)

# Some missed data and discrepansies in endocrine treatment
# (when compare hormone, hormone_self_mra.x and hormone_self_mra.y)
hormone[1:5]
hormone_self_mra.x[1:5]
hormone_self_mra.y[1:5]
sum(hormone_self_mra.x != hormone_self_mra.y, na.rm=TRUE)
sum(hormone != hormone_self_mra.x, na.rm=TRUE)
sum(is.na(hormone))
sum(is.na(hormone_self_mra.x))
sum(is.na(hormone_self_mra.y))

# No missed data or discrepansies in cytotoxic treatment
# (when compare chemo, chemo_self_mra.x and chemo_self_mra.y)
chemo[1:5]
sum(chemo_self_mra.x != chemo_self_mra.y)
sum(chemo != chemo_self_mra.x)

# No missed data or discrepansies in x-ray treatment
# (when compare XRTBrCHAR, XRTBreast.x and XRTBreast.y)
XRTBrCHAR[1:5]
sum(XRTBreast.x != XRTBreast.y)
sum(XRTBrCHAR != XRTBreast.x)

# Eigenvectors
sum(Eigen_1.x != Eigen_1.y)
sum(Eigen_2.x != Eigen_2.y)
sum(Eigen_3.x != Eigen_3.y)
sum(Eigen_4.x != Eigen_4.y)
sum(Eigen_5.x != Eigen_5.y)

# Mixed fields
sum(setno.x != setno.y)
sum(as.vector(registry.x) != as.vector(registry.y))
sum(good_location.x != good_location.y)
sum(status.x != status.y)
sum(status2.x != status2.y)
sum(status.x != status2.y)
sum(race.x != race.y)
sum(sub_race != race.y)
sum(sub_race)

# ---------- Non x-y ---------- #

# X-ray dose prsented in differentways
dose[1:5]
dose_caseloc[1:5]
dose_num[1:5]
avedose[1:5]

horm_tmx[1:5]
tmx[1:5]

cmf[1:5]
CMF[1:5]
sum(as.vector(cmf) != as.vector(CMF))

cmf_012[1:5]
CMF_Only[1:5]

# ========================= Done till here ================================= #

samples.df %>% 
  filter(grepl("not_annotated", filter)) %>% 
  arrange(filter)

not_annotated <-
  samples.df %>% 
  filter(grepl("not_annotated", filter)) %>% 
  select(gwas_id)

not_annotated <- as.vector(not_annotated$gwas_id)

samples.df %>% 
  filter(gwas_id %in% not_annotated)

demographics.df %>% 
  select(labid.x,labid.y) %>% 
  filter(labid.x %in% not_annotated | labid.y %in% not_annotated)

demographics.df %>% 
  select(starts_with("labid")) %>% 
  filter(is.na(labid.x) | is.na(labid.y))

demographics.df %>% 
  select(starts_with("labid")) %>% 
  filter(labid.x %in% wes.done.not.annotated | labid.y %in% wes.done.not.annotated)



sum(chemo != chemo_self_mra)
sum(hormone != hormone_self_mra, na.rm=TRUE)
sum(status2 != status2)
sum(race)
sum(dsmiss)
sum(good_location)
sum(Deleterious)
sum(family_history)
sum(XRTBrCHAR != XRTBreast)
sum(sub_dx_age_cat)

detach(pheno.df)

```

# reshape_demographics



# select_relevant_phenotypes_annotations

```{r select_relevant_phenotype_annotations}

phenotypes.df <- 
  covar.df %>% 
  select(labid, cc, 
         sub_dx_age, offset, family_history, 
         chemo, hormone, XRTBreast, dose, 
         Eigen_1, Eigen_2, Eigen_3, Eigen_4, Eigen_5)

colnames(phenotypes.df)[1] <- "gwas_id" # for merging with samples.df

rm(covar.df)

```



# add_WES_cases_ids_to_phenotypes_table

```{r add_WES_cases_ids_to_phenotypes_table}

# Explore cases submitted for WES
sort(table(samples.df$gwas_id))

# Explore cases annotated in GWAS
sort(table(phenotypes.df$gwas_id))

# Merge samples and phenotypes tables
merged.df <- merge(samples.df, phenotypes.df, by="gwas_id")
dim(merged.df)

# Explore cases in the merged set cases missed and measured twice in wes
sort(table(merged.df$gwas_id))

# Cases without wes results (failes sequencing)
sort(table(merged.df$gwas_id))[ sort(table(merged.df$gwas_id)) == 0 ]

# Cases sequenced twice
sort(table(merged.df$gwas_id))[ sort(table(merged.df$gwas_id)) == 2 ]

# Look closer into two samples that had been measured twice
merged.df[merged.df$gwas_id %in% c("id259643", "id272715"),]

# For case sequenced twice: exclude measurement with lower coverage (as in lab records)
cases_to_exclude <- merged.df$merged_id %in% c("P4_F01_id259643", "P5_D07_id272715")
phenotypes.df <- merged.df[!cases_to_exclude, ]
dim(phenotypes.df)

# Sort phenotypes table
phenotypes.df <- phenotypes.df %>% arrange(wes_id)

# Clean-up
rm(merged.df, samples.df, cases_to_exclude)

```

At this point phenotypes table contains information about 498 samples.  
These are all annotated samples placed on WES plates.  
12 of 512 WES samples were removed because they had no phenotype annotations.
2 samples were removed because they were placed to the plates in duplicate.

# reshape_genotypes_table

```{r reshape_genotypes_table}

# Rename genotype table
genotypes.mx <- gt.pf.msk.lof.mx
rm(gt.pf.msk.lof.mx)

# Update column names in genotype table
colnames(genotypes.mx) <- sub(".GT$","",colnames(genotypes.mx))
colnames(genotypes.mx)

# Get IDs of annotated cases (from GWAS dataset)
annotated.cases <- as.vector(phenotypes.df$wes_id)
length(annotated.cases)

# Get IDs of successfully sequenced cases (from WES genotypes table)
sequenced.cases <- colnames(genotypes.mx)
length(sequenced.cases)

# Cases to analyse
selected.cases <- sort(intersect(annotated.cases, sequenced.cases))
length(selected.cases)
rm(annotated.cases, sequenced.cases)

# Select cases with annotated phenotypes
genotypes.mx <- genotypes.mx[,selected.cases]
dim(genotypes.mx)

# Use merged IDs for column names in genotype table
rownames(phenotypes.df) <- phenotypes.df$wes_id
colnames(genotypes.mx) <- phenotypes.df[selected.cases,]$merged_id

```

# update_phenotypes_table

```{r update_phenotypes_table}

# Remove ceses that failed sequencing
phenotypes.df <- phenotypes.df[selected.cases,]

# Reorder columns
phenotypes.df <- phenotypes.df[,c("merged_id", "wes_id", "gwas_id", "cc", "offset", "family_history", "sub_dx_age", "chemo", "hormone", "XRTBreast", "Eigen_1", "Eigen_2", "Eigen_3", "Eigen_4", "Eigen_5")]
  
# Place merged id to rownames
rownames(phenotypes.df) <- phenotypes.df[,1]

# Clean-up
rm(selected.cases)

```

At this point phenotypes and genotypes tables contains information about 483 samples.  
These are all annotated samples which were successfully sequenced:  
- 2 samples were removed because they were placed to the plates in duplicate
- 12 of 512 WES samples were removed because they had no phenotype annotations
- 15 annotated cases failed sequencing

Variants and genotypes tables contain information about 5,439 high impact variants 
that have a potential to cause loss of the protein functuon.

# save_data

```{r save_data}

save.image(file="data/s04_reshape_data_feb2016.RData")

```

# final_section

```{r final_section}

sessionInfo()
Sys.time()

```
