---
title: "Update cases IDs, cases annotations and file names: wecare feb2016"
author: "Alexey Larionov"
output: html_document
---

started: 01Mar2016
last updated: 15May2016

# start_section

```{r start_section}

# Start time
Sys.time()

# Clean-up
rm(list=ls())
graphics.off()

# Set root working folder
library(knitr)
opts_knit$set(root.dir = "/scratch/medgen/scripts/rscripts_05.16")
#setwd("/scratch/medgen/scripts/rscripts_05.16")

#opts_knit$set(root.dir = "C:\\Users\\larion01\\Documents\\GitHub\\R1")
#setwd("C:\\Users\\larion01\\Documents\\GitHub\\R1")

# Required libraries
library(tidyr)
library(dplyr)
library(stringr)

```

# load_data

```{r load_data}

load(file="data/s04_filter_by_effect_feb2016.RData")
ls()

vv.df <- vv.std
gt.mx <- gt.std

rm(vv.std, gt.std, gq.std, dp.std, ref.std, alt.std)
rm(vv.str, gt.str, gq.str, dp.str, ref.str, alt.str)
rm(vv.rel, gt.rel, gq.rel, dp.rel, ref.rel, alt.rel)

ls()

```

# convert_dfs_to_tables - do it later or not do at all: messes with row manes!!

```{r convert_dfs_to_tables}

covar.tb <- tbl_df(covar.df)
rownames(covar.tb) <- rownames(covar.df)
rm(covar.df)

demographics.tb <- tbl_df(demographics.df)
rownames(demographics.tb) <- rownames(demographics.df)
rm(demographics.df)

samples.tb <- tbl_df(samples.df)
rownames(samples.tb) <- rownames(samples.df)
rm(samples.df)

vv.tb <- tbl_df(vv.df)
rownames(vv.tb) <- rownames(vv.df)
rm(vv.df)

```

# check_data

```{r check_data}

dim(covar.tb)
class(covar.tb)
str(covar.tb)
covar.tb[1:5,1:5]
rownames(covar.tb)[1:5]

dim(demographics.tb)
class(demographics.tb)
str(demographics.tb)
demographics.tb[1:5,1:5]
rownames(demographics.tb)[1:5]

dim(samples.tb)
class(samples.tb)
str(samples.tb)
samples.tb[1:5,]
rownames(samples.tb)[1:5]

dim(vv.tb)
class(vv.tb)
str(vv.tb)
vv.tb[1:5,1:5]
rownames(vv.tb)[1:5]

dim(gt.mx)
class(gt.mx)
gt.mx[1:25,1:5]

# Check consistence of rownames
sum(rownames(gt.mx) != rownames(vv.tb), na.rm=TRUE)

```

# reshape_variants_annotations

```{r reshape_variants_annotations}

# select relevant variants annotations
variants.tb <- 
  vv.tb %>% 
  select(SYMBOL, TYPE, CHROM, POS, REF, ALT, AC, AF, AN, 
         Consequence, SIFT, PolyPhen, CLIN_SIG, 
         GMAF, EUR_MAF, ALT_frequency_in_1k_95)

class(variants.tb)
rownames(variants.tb)[1:5]

rm(vv.tb)

# Explore cases with confusing GMAF/EUR_MAF containing & symbol
nrow(variants.tb %>% filter(grepl("&",GMAF)))
nrow(variants.tb %>% filter(grepl("&",EUR_MAF)))
variants.tb %>% 
  filter(grepl("&",EUR_MAF)) %>% 
  select(c(1:6), GMAF, EUR_MAF)

# There are 3 variants with triplicated values separated by &
variants.tb[c(1235, 11405, 15894),c("GMAF", "EUR_MAF")]

# Split cases with confusing EUR_MAF containing & symbol
variants.tb <- 
  variants.tb %>% 
  separate(EUR_MAF, c("EUR_MAF1", "EUR_MAF2"), "&") 

- Lost row names !

# All duplicated values are the same
sum(variants.tb$EUR_MAF1 != variants.tb$EUR_MAF2, na.rm=TRUE)

# Remove one of the duplicates
variants.tb <- 
  variants.tb %>% 
  select(-EUR_MAF2)

# split GMAF and EUR_MAF1
variants.tb <- 
  variants.tb %>% 
  separate(GMAF, c("GMAF_Allele", "GMAF_Fraction"), ":") %>% 
  separate(EUR_MAF1, c("EUR_MAF_Allele", "EUR_MAF_Fraction"), ":") 

# Convert columns to numeric etc
as.numeric(variants.tb$GMAF_Fraction) -> variants.tb$GMAF_Fraction
as.numeric(variants.tb$EUR_MAF_Fraction) -> variants.tb$EUR_MAF_Fraction
NA -> variants.tb$GMAF_Allele[variants.tb$GMAF_Allele == ""]
NA -> variants.tb$EUR_MAF_Allele[variants.tb$EUR_MAF_Allele == ""]
as.factor(variants.tb$GMAF_Allele) -> variants.tb$GMAF_Allele
as.factor(variants.tb$EUR_MAF_Allele) -> variants.tb$EUR_MAF_Allele

# Check variants table
dim(variants.tb)
class(variants.tb)
str(variants.tb)
variants.tb[1:5,1:5]
rownames(variants.tb)[1:5]

```

# explore_covar

```{r explore_covar}

dim(covar.df)
colnames(covar.df)

attach(covar.df)

sum(chemo != chemo_self_mra)
sum(hormone != hormone_self_mra, na.rm=TRUE)
sum(status2 != status2)
sum(race)
sum(dsmiss)
sum(good_location)
sum(Deleterious)
sum(family_history)
sum(XRTBrCHAR != XRTBreast)
sum(sub_dx_age_cat)

detach(covar.df)

```

# explore_demographics

```{r explore_demographics}

dim(demographics.df)
colnames(demographics.df)

demographics.tb <- tbl_df(demographics.df)
class(demographics.df)
class(demographics.tb)

samples.df %>% 
  filter(grepl("not_annotated", filter)) %>% 
  arrange(filter)

not_annotated <-
  samples.df %>% 
  filter(grepl("not_annotated", filter)) %>% 
  select(gwas_id)

not_annotated <- as.vector(not_annotated$gwas_id)

samples.df %>% 
  filter(gwas_id %in% not_annotated)

demographics.df %>% 
  select(labid.x,labid.y) %>% 
  filter(labid.x %in% not_annotated | labid.y %in% not_annotated)

attach(demographics.df)

demographics.df %>% 
  select(starts_with("labid")) %>% 
  filter(is.na(labid.x) | is.na(labid.y))

demographics.df %>% 
  select(starts_with("labid")) %>% 
  filter(labid.x %in% wes.done.not.annotated | labid.y %in% wes.done.not.annotated)



sum(chemo != chemo_self_mra)
sum(hormone != hormone_self_mra, na.rm=TRUE)
sum(status2 != status2)
sum(race)
sum(dsmiss)
sum(good_location)
sum(Deleterious)
sum(family_history)
sum(XRTBrCHAR != XRTBreast)
sum(sub_dx_age_cat)

detach(demographics.df)

```

# select_relevant_phenotypes_annotations

```{r select_relevant_phenotype_annotations}

phenotypes.df <- 
  covar.df %>% 
  select(labid, cc, 
         sub_dx_age, offset, family_history, 
         chemo, hormone, XRTBreast, dose, 
         Eigen_1, Eigen_2, Eigen_3, Eigen_4, Eigen_5)

colnames(phenotypes.df)[1] <- "gwas_id" # for merging with samples.df

rm(covar.df)

```

# add_WES_cases_ids_to_phenotypes_table

```{r add_WES_cases_ids_to_phenotypes_table}

# Explore cases submitted for WES
sort(table(samples.df$gwas_id))

# Explore cases annotated in GWAS
sort(table(phenotypes.df$gwas_id))

# Merge samples and phenotypes tables
merged.df <- merge(samples.df, phenotypes.df, by="gwas_id")
dim(merged.df)

# Explore cases in the merged set cases missed and measured twice in wes
sort(table(merged.df$gwas_id))

# Cases without wes results (failes sequencing)
sort(table(merged.df$gwas_id))[ sort(table(merged.df$gwas_id)) == 0 ]

# Cases sequenced twice
sort(table(merged.df$gwas_id))[ sort(table(merged.df$gwas_id)) == 2 ]

# Look closer into two samples that had been measured twice
merged.df[merged.df$gwas_id %in% c("id259643", "id272715"),]

# For case sequenced twice: exclude measurement with lower coverage (as in lab records)
cases_to_exclude <- merged.df$merged_id %in% c("P4_F01_id259643", "P5_D07_id272715")
phenotypes.df <- merged.df[!cases_to_exclude, ]
dim(phenotypes.df)

# Sort phenotypes table
phenotypes.df <- phenotypes.df %>% arrange(wes_id)

# Clean-up
rm(merged.df, samples.df, cases_to_exclude)

```

At this point phenotypes table contains information about 498 samples.  
These are all annotated samples placed on WES plates.  
12 of 512 WES samples were removed because they had no phenotype annotations.
2 samples were removed because they were placed to the plates in duplicate.

# reshape_genotypes_table

```{r reshape_genotypes_table}

# Rename genotype table
genotypes.mx <- gt.pf.msk.lof.mx
rm(gt.pf.msk.lof.mx)

# Update column names in genotype table
colnames(genotypes.mx) <- sub(".GT$","",colnames(genotypes.mx))
colnames(genotypes.mx)

# Get IDs of annotated cases (from GWAS dataset)
annotated.cases <- as.vector(phenotypes.df$wes_id)
length(annotated.cases)

# Get IDs of successfully sequenced cases (from WES genotypes table)
sequenced.cases <- colnames(genotypes.mx)
length(sequenced.cases)

# Cases to analyse
selected.cases <- sort(intersect(annotated.cases, sequenced.cases))
length(selected.cases)
rm(annotated.cases, sequenced.cases)

# Select cases with annotated phenotypes
genotypes.mx <- genotypes.mx[,selected.cases]
dim(genotypes.mx)

# Use merged IDs for column names in genotype table
rownames(phenotypes.df) <- phenotypes.df$wes_id
colnames(genotypes.mx) <- phenotypes.df[selected.cases,]$merged_id

```

# update_phenotypes_table

```{r update_phenotypes_table}

# Remove ceses that failed sequencing
phenotypes.df <- phenotypes.df[selected.cases,]

# Reorder columns
phenotypes.df <- phenotypes.df[,c("merged_id", "wes_id", "gwas_id", "cc", "offset", "family_history", "sub_dx_age", "chemo", "hormone", "XRTBreast", "Eigen_1", "Eigen_2", "Eigen_3", "Eigen_4", "Eigen_5")]
  
# Place merged id to rownames
rownames(phenotypes.df) <- phenotypes.df[,1]

# Clean-up
rm(selected.cases)

```

At this point phenotypes and genotypes tables contains information about 483 samples.  
These are all annotated samples which were successfully sequenced:  
- 2 samples were removed because they were placed to the plates in duplicate
- 12 of 512 WES samples were removed because they had no phenotype annotations
- 15 annotated cases failed sequencing

Variants and genotypes tables contain information about 5,439 high impact variants 
that have a potential to cause loss of the protein functuon.

# save_data

```{r save_data}

save.image(file="data/s04_reshape_data_feb2016.RData")

```

# final_section

```{r final_section}

sessionInfo()
Sys.time()

```
